csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -759,6 +759,312 @@ std::unique_ptr<ReductionParams> innerOuterPersistentHeuristic(
+  // and read of the partial outer reduction result.
+  // For write to tmp gmem, follows vectorization factor of inner reduction
+  //                        but don't exceed 16 bytes.
+  // For read from tmp gmem, since the paralelization is changed, a different
+  //                         vectorization factor is used to optimize the
+  //                         number of reaductions per thread.
+  auto getOuterReductionBufferVectFactor = [&](int64_t inner_vect) {
+    constexpr int64_t max_gmem_vect_access_bytes = 16;
+    const int64_t max_tmp_gmem_vect_factor = std::min(
+        max_gmem_vect_access_bytes / (int64_t)tmp_gmem_dtype_size, inner_vect);
+    int64_t tmp_gmem_write_vect = max_tmp_gmem_vect_factor;
+    const int64_t workload_per_thread = inner_dim_numel >= 4096 ? 4l : 2l;
+    int64_t vectorization_factor_outer =
+        std::min(workload_per_thread, max_tmp_gmem_vect_factor);
+    return std::make_pair(tmp_gmem_write_vect, vectorization_factor_outer);
+  };
+
+  // In the outer reduction part of the kernel, inner and outer dims are
+  // parallelized as:
+  // --- inner dim: vect, bdimx, gdimy ----
+  // --- outer dim: bdimy -----------------
+  // This function splits the threads_per_block into bdimx and bdimy using:
+  // bdimx = ceilDiv(inner_dim_numel / vect, gdimy)
+  // bdimy = threads_per_block / bdimx
+  auto getBdimxBdimy = [&](int64_t threads_per_block,
+                           int64_t vectorization_factor_outer,
+                           int64_t gdimy) {
+    // For widely used hidden sizes, threads_per_block has factor of 8, roundup
+    // to increase the probability of bdimx * bdimy == threads_per_block.
+    int64_t bdimx = scheduler_utils::roundUpPow2Or8(
+        ceilDiv(inner_dim_numel / vectorization_factor_outer, gdimy));
+    // if still not divisible, e.g. threads_per_block = 256, bdimx = 40.
+    // increase bdimx to make it divisible. Under worst case, bdimx equals to
+    // threads_per_block.
+    while (threads_per_block % bdimx) {
+      bdimx = std::min(bdimx + 8, threads_per_block);
+    }
+    // Set OuterParams Reduction dim: bdimy.
+    int64_t bdimy = threads_per_block / bdimx;
+    NVF_ERROR(
+        bdimy * bdimx == threads_per_block,
+        " threads_per_block must be divisible by bdimx and bdimy.");
+    return std::make_pair(bdimx, bdimy);
+  };
+
+  // Get the heuristics given vectorization factor and threads per block
+  auto getHeuristicsGivenVectThreads = [&](int64_t vect_factor,
+                                           int64_t threads_per_block) {
+    InnerOuterParams iop;
+    // (1) inner reduction
+    // Reduction dim: inner_batch, threads_per_block, vect_factor
+    // Iteration dim: gdimy
+    iop.inner_vect = vect_factor;
+    iop.threads_per_block = threads_per_block;
+    iop.inner_batch =
+        ceilDiv(inner_dim_numel / iop.inner_vect, iop.threads_per_block);
+    iop.gdimy = device_multiprocessor_count;
```

Comment:
Instead of computing the max possible `gdimy` using `getGdimy()`, `gdimy` is fixed to SM count.

Reply:
other than this change, I also removed special heuristics for `inner_dim_numel <= 1024`, which is not required for the new scheduler. Other parts of this heuristics function will be cleaned in following PRs.

Reply:
Thanks for pointing this one out, otherwise, it's tricky to get the changes..